File: json2html.js
const fs = require('fs');

const input = process.argv[2];
const output = process.argv[3];

if (!input || !output) {
  console.error("Usage: node json2html.js <input.json> <output.html>");
  process.exit(1);
}

const data = JSON.parse(fs.readFileSync(input, 'utf8'));
const vulnerabilities = data.vulnerabilities || [];
const components = data.components || [];

function getComponent(ref) {
  return components.find(c => c['bom-ref'] === ref);
}

function getSeverityBadge(sev) {
  const s = (sev || '').toLowerCase();
  if (s === 'critical') return '<span class="badge bg-danger">Critical'</span>;
  if (s === 'high') return '<span class="badge bg-warning text-dark">High'</span>;
  if (s === 'medium') return '<span class="badge bg-info text-dark">Medium'</span>;
  if (s === 'low') return '<span class="badge bg-success">Low'</span>;
  return '<span class="badge bg-secondary">Unknown'</span>;
}

const severityCount = {
  Critical: 0,
  High: 0,
  Medium: 0,
  Low: 0,
  Unknown: 0
};

const rows = vulnerabilities.flatMap((vuln, index) => {
  const id = vuln.id || 'N/A';
  const severityRaw = vuln.ratings?.[0]?.severity || 'Unknown';
  const severity = severityRaw.charAt(0).toUpperCase() + severityRaw.slice(1).toLowerCase();
  const score = vuln.ratings?.[0]?.score ?? 'N/A';
  const desc = vuln.description || 'N/A';
  const sourceUrl = vuln.source?.url || 'N/A';
  const affects = vuln.affects || [];

  if (severityCount[severity] !== undefined) severityCount[severity]++;
  else severityCount.Unknown++;

  return affects.map((aff, subIndex) => {
    const component = getComponent(aff.ref);
    const compName = component ? `${component.name}@${component.version}` : aff.ref;
    const filePath = component?.properties?.find(p => p.name.includes("location") && p.value)?.value || 'N/A';

    return `
      <tr>
        <td>${index + 1}${affects.length > 1 ? `.${subIndex + 1}` : ''}</td>
        <td>${id}</td>
        <td>${severity}</td>
        <td>${score}</td>
        <td>${compName}</td>
        <td>${filePath}</td>
        <td><a href="${sourceUrl}" target="_blank">${sourceUrl}</a></td>
        <td>${desc}</td>
      </tr>
    `;
  });
}).join('\n');

const html = `
<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8" />
  <title>SBOM Vulnerability Report</title>
  <link href="https://cdn.jsdelivr.net/npm/bootstrap@5.3.3/dist/css/bootstrap.min.css" rel="stylesheet">
  <link href="https://cdn.datatables.net/1.13.6/css/dataTables.bootstrap5.min.css" rel="stylesheet">
  <style>
    body { padding: 20px; background-color: #f8f9fa; }
    .badge { font-size: 0.9em; }
    .filter-container { margin-bottom: 20px; }
    .filter-container label { margin-right: 10px; }
    .summary-badges span { margin-right: 10px; }
  </style>
</head>
<body>
  <div class="container-fluid">
    <h2 class="mb-4">SBOM Vulnerability Report</h2>

    <div class="summary-badges mb-2">
      <strong>Total Vulnerabilities:</strong> ${vulnerabilities.length}
      <span class="badge bg-danger">Critical: ${severityCount.Critical}</span>
      <span class="badge bg-warning text-dark">High: ${severityCount.High}</span>
      <span class="badge bg-info text-dark">Medium: ${severityCount.Medium}</span>
      <span class="badge bg-success">Low: ${severityCount.Low}</span>
      <span class="badge bg-secondary">Unknown: ${severityCount.Unknown}</span>
    </div>

    <div class="filter-container mb-3">
      <label for="severityFilter"><strong>Filter by Severity:</strong></label>
      <select id="severityFilter" class="form-select w-auto d-inline-block">
        <option value="">All</option>
        <option value="Critical">Critical</option>
        <option value="High">High</option>
        <option value="Medium">Medium</option>
        <option value="Low">Low</option>
        <option value="Unknown">Unknown</option>
      </select>
    </div>

    <div class="table-responsive">
      <table id="vulnTable" class="table table-striped table-bordered">
        <thead class="table-dark">
          <tr>
            <th>#</th>
            <th>Vuln ID</th>
            <th>Severity</th>
            <th>CVSS Score</th>
            <th>Affected Component</th>
            <th>File Path</th>
            <th>Source URL</th>
            <th>Description</th>
          </tr>
        </thead>
        <tbody>
          ${rows || `<tr><td colspan="8" class="text-center">No vulnerabilities found.</td></tr>`}
        </tbody>
      </table>
    </div>
  </div>

  <script src="https://code.jquery.com/jquery-3.7.0.min.js"></script>
  <script src="https://cdn.jsdelivr.net/npm/bootstrap@5.3.3/dist/js/bootstrap.bundle.min.js"></script>
  <script src="https://cdn.datatables.net/1.13.6/js/jquery.dataTables.min.js"></script>
  <script src="https://cdn.datatables.net/1.13.6/js/dataTables.bootstrap5.min.js"></script>
  <script>
    $(document).ready(function() {
      const table = $('#vulnTable').DataTable({
        responsive: true,
        pageLength: 10,
        lengthMenu: [10, 25, 50, 100],
        language: { search: "Search:" },
        columnDefs: [
          {
            targets: 2,
            render: function(data, type, row) {
              const s = data.toLowerCase();
              if (s === 'critical') return '<span class="badge bg-danger">Critical';</span>
              if (s === 'high') return '<span class="badge bg-warning text-dark">High';</span>
              if (s === 'medium') return '<span class="badge bg-info text-dark">Medium';</span>
              if (s === 'low') return '<span class="badge bg-success">Low';</span>
              return '<span class="badge bg-secondary">Unknown';</span>
            }
          }
        ]
      });

      // Custom severity filter
      $('#severityFilter').on('change', function() {
        const selected = $(this).val();
        if (selected) {
          table.column(2).search('^' + selected + '$', true, false).draw();
        } else {
          table.column(2).search('').draw();
        }
      });
    });
  </script>
</body>
</html>
`;

fs.writeFileSync(output, html, 'utf8');
console.log(`✅ Filterable report generated at ${output}`);
 

File: json2html.py
#!/usr/bin/env python3
"""
Script to convert a vulnerability JSON file to a structured HTML report
"""

import json
import sys
from collections import Counter

def parse_vuln_json(file_path):
    try:
        with open(file_path, 'r', encoding='utf-8') as f:
            data = json.load(f)
    except Exception as e:
        print(f"Error reading JSON: {e}")
        sys.exit(1)

    results = []
    artifact_name = data.get('ArtifactName', '')
    created_at = data.get('CreatedAt', '')

    for result in data.get('Results', []):
        target = result.get('Target', '')
        for vuln in result.get('Vulnerabilities', []):
            results.append({
                'Artifact Name': artifact_name,
                'Target': target,
                'Vulnerability ID': vuln.get('VulnerabilityID', ''),
                'CWE IDs': ', '.join(vuln.get('CweIDs', [])),
                'Severity': vuln.get('Severity', '').upper(),
                'Severity Source': vuln.get('SeveritySource', ''),
                'Package ID': vuln.get('PkgID', ''),
                'Package Name': vuln.get('PkgName', ''),
                'Title': vuln.get('Title', ''),
                'Description': vuln.get('Description', ''),
                'Installed Version': vuln.get('InstalledVersion', ''),
                'Fixed Version': vuln.get('FixedVersion', ''),
                'Primary URL': vuln.get('PrimaryURL', ''),
                'Published Date': vuln.get('PublishedDate', ''),
                'Last Modified Date': vuln.get('LastModifiedDate', '')
            })

    return results

def generate_html(vulnerabilities, output_file):
    if not vulnerabilities:
        print("No data to write to HTML.")
        return

    severity_counts = Counter(v['Severity'] for v in vulnerabilities)
    total = len(vulnerabilities)

    html = f"""<!DOCTYPE html>
<html lang="en">
<head>
<meta charset="UTF-8">
<title>Vulnerability Report</title>
<style>
    body {{ font-family: Arial, sans-serif; padding: 20px; }}
    h1, h2 {{ color: #333; }}
    table {{ border-collapse: collapse; width: 100%; margin-top: 20px; }}
    th, td {{ border: 1px solid #ccc; padding: 8px; text-align: left; }}
    th {{ background-color: #f2f2f2; }}
    tr:nth-child(even) {{ background-color: #fafafa; }}
    .critical {{ background-color: #ff9999; }}
    .high {{ background-color: #ffc7ce; }}
    .medium {{ background-color: #ffeb9c; }}
    .low {{ background-color: #c6efce; }}

    .button-bar button {{
        margin-right: 10px;
        padding: 8px 12px;
        border: none;
        cursor: pointer;
        font-weight: bold;
    }}
    .filter-active {{ background-color: #007bff; color: white; }}
</style>
<script>
function filterSeverity(level) {{
    const rows = document.querySelectorAll('table#vulnTable tbody tr');
    rows.forEach(row => {{
        const severity = row.getAttribute('data-severity');
        row.style.display = (level === 'ALL' || severity === level) ? '' : 'none';
    }});

    document.querySelectorAll('.button-bar button').forEach(btn => btn.classList.remove('filter-active'));
    document.getElementById('btn-' + level).classList.add('filter-active');
}}
</script>
</head>
<body>
<h1>Vulnerability Report</h1>
<p><strong>Total Vulnerabilities:</strong> {total}</p>

<h2>Severity Summary</h2>
<table>
<tr>
  <th>CRITICAL</th><th>HIGH</th><th>MEDIUM</th><th>LOW</th>
</tr>
<tr>
  <td>{severity_counts.get('CRITICAL', 0)}</td>
  <td>{severity_counts.get('HIGH', 0)}</td>
  <td>{severity_counts.get('MEDIUM', 0)}</td>
  <td>{severity_counts.get('LOW', 0)}</td>
</tr>
</table>

<div class="button-bar" style="margin-top: 20px;">
  <button id="btn-ALL" class="filter-active" onclick="filterSeverity('ALL')">Show All</button>
  <button id="btn-CRITICAL" onclick="filterSeverity('CRITICAL')">Critical</button>
  <button id="btn-HIGH" onclick="filterSeverity('HIGH')">High</button>
  <button id="btn-MEDIUM" onclick="filterSeverity('MEDIUM')">Medium</button>
  <button id="btn-LOW" onclick="filterSeverity('LOW')">Low</button>
</div>

<h2>Detailed Vulnerabilities</h2>
<table id="vulnTable">
<thead><tr>
<th>Sr. No.</th>"""

    headers = list(vulnerabilities[0].keys())
    for header in headers:
        html += f"<th>{header}</th>"
    html += "</tr></thead><tbody>\n"

    for i, vuln in enumerate(vulnerabilities, start=1):
        severity_class = vuln.get("Severity", "").lower()
        html += f'<tr class="{severity_class}" data-severity="{vuln.get("Severity", "")}">'
        html += f"<td>{i}</td>"
        for header in headers:
            value = vuln.get(header, "")
            if header == "Primary URL" and value:
                value = f'<a href="{value}" target="_blank">{value}</a>'
            html += f"<td>{value}</td>"
        html += "</tr>\n"

    html += """
</tbody></table>
</body>
</html>"""

    with open(output_file, 'w', encoding='utf-8') as f:
        f.write(html)

    print(f"HTML report created successfully: {output_file}")

def main():
    if len(sys.argv) != 3:
        print("Usage: python3 json2html.py input.json output.html")
        sys.exit(1)

    input_file = sys.argv[1]
    output_file = sys.argv[2]

    vulnerabilities = parse_vuln_json(input_file)
    generate_html(vulnerabilities, output_file)

if __name__ == "__main__":
    main()

